<div style="border:solid green 2px; padding: 20px">
<b>Hello, Anton!</b>
    
Grigoriy is here again. This is our final metting :)
    
You're all data analysts now, so no yellow comments here. In my feedback I tried to give the comments in the way your future team lead would do.    

You can find my comments in <font color='green'>green</font> or <font color='red'>red</font> boxes like this:

<div class="alert alert-success">
<b>Reviewer's comment:</b> if everything is done succesfully
</div>

<div class="alert alert-danger">
<b>Reviewer's comment:</b> if the block requires some corrections. Work can't be accepted with the red comments.
</div>
    
Let's work on the project in dialogue: if you change something in the project or respond to my comments, write about it. It will be easier for me to track the changes if you highlight your comments:
    
<div class="alert alert-info"> <b>Student's comment:</b> For example like this.</div>    
    
      
<div>

### <font color='orange'>General feedback</font>
Thank you for sending your project! Excelent job, all queries are correct. Congratulations, your project has been accepted! Good luck in the final task 💪 

---

### Final project. SQL

This is work of Anton Rubenchik anton.rubenchik@gmail.com

##### Table of contents
0. [Introduction](#introduction)<br />
1. [Download the data and prepare it for analysis.](#download)<br />
2. [Step 2. Study the tables.](#tables)<br />
2.1 [Table "books"](#books)<br />
2.2 [Table "authors"](#authors)<br />
2.3 [Table "publishers"](#publishers)<br />
2.4 [Table "ratings"](#ratings)<br />
2.5 [Table "reviews"](#reviews)<br />
3. [Step 3. Tasks](#tasks)<br />
3.1 [Number of books published after January 1, 2000](#tasks1)<br />
3.2 [Find out the number of user reviews and the average rating for each book.](#tasks2)<br />
3.3 [Identify the publisher with the most books with more than 50 pages.](#tasks3)<br />
3.4 [Identify the author with the highest average book rating.](#tasks4)<br />
3.4.1 [The author of the book with the highest rating.](#tasks4_1)<br />
3.4.2 [Author of top-rated books.](#tasks4_2)<br />
3.5 [Find the average number of text reviews among users who rated more than 50 books](#tasks5)


### Introduction<br /><a name="introduction"></a>
Task:<br />
The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were at home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers. <br /><br />
We've been got a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

We need to analyze the data and get answers to the following questions:
1. Number of books published after January 1, 2000
2. Find out the number of user reviews and the average rating for each book.
3. Identify the publisher with the most books with more than 50 pages.
4. Identify the author with the highest average book rating.
5. Find the average number of text reviews among users who rated more than 50 books. 

<div class="alert alert-success">
<b>Reviewer's comment:</b> Great start!
</div>

### Step 1. Download the data and prepare it for analysis.<a name="download"></a>

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine

In [2]:
db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

##### Conclusion: <br />
We received data from the server, let's examine the data in the tables.

### Step 2. Study the tables.<a name="tables"></a>

#### 2.1 Table "books"<a name="books"></a>

In [3]:
query = """
SELECT *
FROM books
limit 5
"""

In [4]:
results = pd.io.sql.read_sql(query, con = engine)

display(results)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

#### 2.2 Table "authors"<a name="authors"></a>

In [5]:
query = """
SELECT *
FROM authors
limit 5
"""

In [6]:
results = pd.io.sql.read_sql(query, con = engine)

display(results)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

#### 2.3 Table "publishers"<a name="publishers"></a>

In [7]:
query = """
SELECT *
FROM publishers
limit 5
"""

In [8]:
results = pd.io.sql.read_sql(query, con = engine)

display(results)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

#### 2.4 Table "ratings"<a name="ratings"></a>

In [9]:
query = """
SELECT *
FROM ratings
limit 5
"""

In [10]:
results = pd.io.sql.read_sql(query, con = engine)

display(results)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

#### 2.5 Table "reviews"<a name="reviews"></a>

In [11]:
query = """
SELECT *
FROM reviews
limit 5
"""

In [12]:
results = pd.io.sql.read_sql(query, con = engine)

display(results)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<div class="alert alert-success">
<b>Reviewer's comment:</b> Well done, let's move on to the queries part.
</div>

### Step 3. Tasks<a name="tasks"></a>

#### 3.1 Number of books published after January 1, 2000<a name="tasks1"></a>

In [13]:
query_task1 = """
-- Find the number of books released after January 1, 2000.
SELECT COUNT(*) as count_books
FROM books
WHERE publication_date > '2000-01-01';  
"""
results = pd.io.sql.read_sql(query_task1, con = engine)
display(results)

count_books
0          819

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

##### Conclusion: <br />
After performing the query, we found out that 819 books were published after 01/01/2000

#### 3.2 Find out the number of user reviews and the average rating for each book.<a name="tasks2"></a>

In [14]:
query_task2 = """
-- Find the number of user reviews and the average rating for each book.
SELECT books.book_id, 
    authors.author,
    books.title, 
    COUNT(distinct review_id) as number_users_reviews, 
    ROUND(AVG(ratings.rating),2) as average_rating
FROM books
    left join reviews on books.book_id = reviews.book_id 
    left join ratings on books.book_id = ratings.book_id 
    left join authors on books.author_id = authors.author_id 
group by books.book_id, authors.author_id
order by average_rating desc 
"""
results = pd.io.sql.read_sql(query_task2, con = engine)
display(results[:5])

book_id                                             author  \
0       86                                    Mercedes Lackey   
1      901  Robert Kirkman/Tony Moore/Charlie Adlard/Cliff...   
2      390                                   William Faulkner   
3      972                                     Jon Kabat-Zinn   
4      136                       John Eldredge/Stasi Eldredge   

                                               title  number_users_reviews  \
0      Arrows of the Queen (Heralds of Valdemar  #1)                     2   
1  The Walking Dead  Book One (The Walking Dead #...                     2   
2                                    Light in August                     2   
3  Wherever You Go  There You Are: Mindfulness Me...                     2   
4  Captivating: Unveiling the Mystery of a Woman'...                     2   

   average_rating  
0             5.0  
1             5.0  
2             5.0  
3             5.0  
4             5.0

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

##### Conclusion: <br />
We processed the data and formed a table in which we counted the number of reviews for each book and the average book rating.

#### 3.3 Identify the publisher with the most books with more than 50 pages.<a name="tasks3"></a>

In [15]:
query_task3 = """
-- Identify the publisher that has released the greatest number of books with more than 50 pages
select 
    distinct publishers.publisher_id 
    , publishers.publisher
    , COUNT(books.book_id) OVER(PARTITION BY publishers.publisher_id) AS count_books 
from books 
    inner join publishers using(publisher_id)
where books.num_pages > 50
order by count_books desc
limit 1
"""
results = pd.io.sql.read_sql(query_task3, con = engine)
display(results[:5])

publisher_id      publisher  count_books
0           212  Penguin Books           42

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

##### Conclusion: <br />
After performing the query, we found out that the Penguin Books publisher has released the most books.

#### 3.4 Identify the author with the highest average book rating.<a name="tasks4"></a>

##### 3.4.1 The author of the book with the highest rating.<a name="tasks4_1"></a>

In [16]:
query_task4 = """
-- Identify the author with the highest average book rating (look only at books with at least 50 ratings).
with books_rating(book_id,title,author,count_marks,avg_rating)
as (
select 
   distinct book_id
   , books.title 
   , authors.author 
   , count(rating_id) over (partition by book_id) as count_marks
   , round(avg(rating) over (partition by book_id),2) as avg_rating
from books 
    left join ratings using(book_id)
    left join authors using(author_id)
)
select 
    book_id
    , title
    , author
    , avg_rating
from books_rating
where 
    (count_marks > 50) and avg_rating = 
       (
        select avg_rating 
        from books_rating
        where count_marks > 50
        order by avg_rating desc
        limit 1
        )
"""
results = pd.io.sql.read_sql(query_task4, con = engine)
display(results[:5])

book_id                                              title  \
0      302  Harry Potter and the Prisoner of Azkaban (Harr...   

                       author  avg_rating  
0  J.K. Rowling/Mary GrandPré        4.41

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

##### Conclusion: <br />
After performing the query, we found out that J.K. Rowling is the author of the top-rated book itself.

##### 3.4.2 Author of top-rated books.<a name="tasks4_2"></a>

In [17]:
query_task4_1 = """
-- Identify the author with the highest average book rating (look only at books with at least 50 ratings).
with books_rating(book_id,author_id, count_marks,avg_rating)
as (
select 
   distinct book_id
   , authors.author_id 
   , count(rating_id) over (partition by book_id) as count_marks
   , round(avg(rating) over (partition by book_id),2) as avg_rating
from books 
   left join ratings using(book_id)
   right join authors using(author_id)
)
select
    authors.author_id
    , authors.author 
    , COUNT(book_id) as count_books
    , AVG(avg_rating) as avg_rat    
from books_rating
inner join authors on books_rating.author_id = authors.author_id 
where count_marks > 50
group by authors.author_id
order by avg_rat desc 
limit 1
"""
results = pd.io.sql.read_sql(query_task4_1, con = engine)
display(results[:5])

author_id                      author  count_books  avg_rat
0        236  J.K. Rowling/Mary GrandPré            4    4.285

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

##### Conclusion: <br />
After completing the request, we found out that J.K. Rowling is the author whose books have the highest ratings.

#### 3.5 Find the average number of text reviews among users who rated more than 50 books.<a name="tasks5"></a>

In [18]:
query_task5 = """
-- Task 5. Find the average number of text reviews among users who rated more than 50 books. 
with users_rated_50_plus (username)
as (
select 
   ratings.username 
from ratings 
group by username
having count(rating_id) > 50
),
users_count_reviews (username, count_reviews)
as (
    SELECT 
        distinct username
        , COUNT(review_id) over (partition by username) as count_reviews
    from users_rated_50_plus
        inner join reviews USING(username)
)
select ROUND(AVG(count_reviews),2) as AVG_REVIEWS 
from users_count_reviews
"""
results = pd.io.sql.read_sql(query_task5, con = engine)
display(results[:5])

avg_reviews
0        24.33

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

##### Conclusion: <br />
After performing the query, we found out that, on average, users make 24.33 reviews.